### PR TITLE
Avoid deep copy on lz4 decompression

### PR DESCRIPTION
**Pull request by crusaderky**
Speed up deserialization when
a. lz4 is installed, and
b. the buffer is compressible, and
c. the buffer is smaller than 64 MiB (`distributed.comm.shard`)
Note that the default chunk size in dask.array is 128 MiB.

Note that this does not prevent a memory flare, as there's an unnecessary deep copy upstream as well:
https://github.com/python-lz4/python-lz4/blob/79370987909663d4e6ef743762768ebf970a2383/lz4/block/_block.c#L256

### CODE REVIEW

File: distributed/protocol/tests/test_protocol.py [hunk 1]
```diff
@@ -313,10 +335,9 @@ def test_dumps_loads_Serialized():
 
 def test_maybe_compress_memoryviews():
     np = pytest.importorskip("numpy")
-    pytest.importorskip("lz4")
     x = np.arange(1000000, dtype="int64")
     compression, payload = maybe_compress(x.data)
-    assert compression == "lz4"
+    assert compression in {"lz4", "snappy", "zstd", "zlib"}
```
Comment (mrocklin) [inline distributed/protocol/tests/test_protocol.py:340]: I think I would be sad if we used `zlib` by default in any configuration.  I'll bet that it's faster to just send data uncompressed over the network.

File: distributed/protocol/tests/test_numpy.py [hunk 2]
```diff
@@ -217,7 +217,6 @@ def test_itemsize(dt, size):
 
 
 def test_compress_numpy():
-    pytest.importorskip("lz4")
```
Comment (mrocklin) [inline distributed/protocol/tests/test_numpy.py:220]: I'm curious, why this change?  If we didn't have lz4, snappy, or zstandard installed (all of which are optional I think) then I'd expect this to fail.  

The only compressor we have by default,  I think, is zlib, and we don't compress with that by default.

**Review by mrocklin — COMMENTED** [review]
In principle this seems fine.  I did have a couple of small questions though.

File: distributed/protocol/tests/test_protocol.py (hunk 1 above)
Comment (crusaderky) [reply distributed/protocol/tests/test_protocol.py]: Ah, you're right. I misread the code; default compression is lz4 -> snappy -> None.
I've amended the tests and added a specific test for the priority order.

File: distributed/protocol/tests/test_numpy.py (hunk 2 above)
Comment (crusaderky) [reply distributed/protocol/tests/test_numpy.py:220]: Actually, if you have snappy but not lz4 it will succeed.
zstandard does not install itself as a default compressor.
Amended the tests to reflect this.
